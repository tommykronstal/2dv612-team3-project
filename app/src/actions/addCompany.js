import { UPDATE_COMPANY_NAME } from './types';

export function updateCompanyName (companyName) {
  return {
      type: UPDATE_COMPANY_NAME,
      companyName: companyName
  };
}

