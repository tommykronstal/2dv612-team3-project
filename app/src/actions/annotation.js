import {UPDATE_ANNOTATION} from './types'

export const updateAnnotation = (annotationText, materialId) => ({
  type: UPDATE_ANNOTATION,
  annotationText,
  materialId
})
