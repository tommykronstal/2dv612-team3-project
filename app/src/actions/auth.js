import { SET_LOGGED_IN, LOGOUT } from './types'
import { getPayloadFromJwtToken } from '../lib/jwt'

export const setUserToLoggedIn = response => ({ 
  type: SET_LOGGED_IN, 
  token: response.token, 
  payload: getPayloadFromJwtToken(response.token) 
})

export const logout = () => ({ type: LOGOUT })
