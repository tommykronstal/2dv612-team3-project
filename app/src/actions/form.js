import {CLEAR_FORM, SUBMIT_FORM, UPDATE_FIELD} from './types'

export const submitForm = (form, endpoint, action, tokenRequire = true, role = '') => ({
  type: SUBMIT_FORM,
  form,
  endpoint,
  action,
  tokenRequire,
  role
})

export const clearForm = form => ({
  type: CLEAR_FORM,
  form
})

export const updateField = (form, field, value) => ({
  type: UPDATE_FIELD,
  form,
  field,
  value
})
