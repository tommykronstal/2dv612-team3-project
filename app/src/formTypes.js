export const LOGIN = 'LOGIN'
export const ADD_COMPANY = 'ADD_COMPANY'
export const ADD_COMPANY_REP = 'ADD_COMPANY_REP'
export const REGISTER_CONSUMER = 'REGISTER_CONSUMER'
export const ADD_PRODUCT = 'ADD_PRODUCT'
export const UPDATE_PRODUCT = 'UPDATE_PRODUCT'
export const ADD_CATEGORY = 'ADD_CATEGORY'