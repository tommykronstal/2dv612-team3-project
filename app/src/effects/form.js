import {put, takeEvery, select, call} from 'redux-saga/effects'
import {SUBMIT_FORM, TOGGLE_LOADING, SET_STATUS} from '../actions/types'
import {clearForm} from '../actions/form'
import {post} from '../lib/http'
import {UPDATE_PRODUCT} from '../formTypes'

export function* watchFormActions() {
  yield takeEvery(SUBMIT_FORM, formRequest)
}

export function* formRequest({endpoint, form, action, tokenRequired}) {
  yield put({type: TOGGLE_LOADING})


  const {token, payload} = yield select(state => ({
    token: state.auth.jwt,
    payload: {
      ...state.form[form],
    },
  }))

  const tokenHeader = {...(token && tokenRequired && {Authorization: token})}

  if (form === UPDATE_PRODUCT) {
    yield sendProductFile(endpoint, token, payload, tokenHeader, action, form)
    return
  }

  console.log(payload)

  const response = yield call(post, endpoint, {
    headers: {
      ...tokenHeader,
    },
    body: JSON.stringify(payload)
  })

  console.log('response', response)

  yield handleResponse(response, form, action)
}

function* handleResponse(response, form, action) {
  if (response.status === 200 || response.status === 201) {
    if (action) {
      yield put(action(response))
    }
    yield put(clearForm(form))
  } else {
    if (response.message)
      yield put({type: SET_STATUS, warning: true, message: response.message})
  }

  yield put({type: TOGGLE_LOADING})
}


/**
 * Used to send data as a form.
 * Need to send it like this since original function to submit forms
 * assumes that it's JSON. (Maybe needs some tweaking in the future...)
 */
function* sendProductFile(endpoint, token, payload, tokenHeader = {}, action, form) {
  const formData = new FormData()

  console.log(payload)

  for( let key in payload) {
    formData.append(key, payload[key])
  }

  const response = yield call(post, endpoint, {
    headers: {
      ...tokenHeader
    },
    isJsonPayload: false,
    body: formData
  })

  yield handleResponse(response, form, action)
}