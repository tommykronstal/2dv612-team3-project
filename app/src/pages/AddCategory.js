import React from 'react';
import {connect} from 'react-redux';

import {ADD_CATEGORY} from '../formTypes';
import {updateField, submitForm} from '../actions/form'
import {setStatus} from '../actions/status'
import FormSection from '../components/common/FormSection'
import Title from '../components/common/Title'
import Input from '../components/common/Input'
import Button from '../components/common/Button'

export const AddCategory = props => (
  <FormSection onSubmit={props.addCategory}>
    <Title>Add a new category</Title>
    <Input
      value={props.form.categoryName}
      type='text'
      name='categoryName'
      label='Category Name'
      onChange={props.updateField}
    />
    <Button primary loading={props.loading.isLoading}>
      Add
    </Button>
  </FormSection>
)

export const handleResponse = _ => setStatus('Category Added')

export const mapStateToProps = ({ loading, form }) => ({
  loading,
  form: form[ADD_CATEGORY] || {}
})

export const mapDispatchToProps = dispatch => ({
  addCategory: () => dispatch(submitForm(
    ADD_CATEGORY,
    '/api/admin/category',
    handleResponse
  )),
  updateField: ({ target }) => dispatch(updateField(ADD_CATEGORY, target.name, target.value))
})

export default connect(mapStateToProps, mapDispatchToProps)(AddCategory)