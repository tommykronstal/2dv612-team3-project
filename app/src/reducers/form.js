import { CLEAR_FORM, UPDATE_FIELD } from '../actions/types'
import { REGISTER_CONSUMER } from '../formTypes'
import USER_TYPES from '../userTypes'

const initialState = {
  [REGISTER_CONSUMER]: {
    role: USER_TYPES.USER
  }
}

export default function status (state = initialState, action = {}) {
  switch (action.type) {
  case CLEAR_FORM:
    return {
      ...state,
      [action.form]: initialState[action.form] || {}
    }
  case UPDATE_FIELD:
    return {
      ...state,
      [action.form]: {
        ...state[action.form],
        [action.field]: action.value
      }
    }
  default:
    return state
  }
}
