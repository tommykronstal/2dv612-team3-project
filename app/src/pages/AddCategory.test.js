import React from 'react'
import renderer from 'react-test-renderer'

// todo not sure how to use mapStateToProps in test
import { AddCategory, mapStateToProps, mapDispatchToProps, handleResponse } from './AddCategory'
import { ADD_CATEGORY } from '../formTypes'
import { submitForm, updateField } from '../actions/form'

describe('AddCategory', () => {

  describe('component', () => {
    it('renders correctly, initial', () => {
      const props = {
        addCategory: jest.fn(),
        updateField: jest.fn(),
        form: {},
        loading: false
      }

      const tree = renderer.create(<AddCategory {...props} />).toJSON()

      expect(tree).toMatchSnapshot()
    })

    it('renders correctly, with data', () => {
      const props = {
        addCompany: jest.fn(),
        updateField: jest.fn(),
        form: {
          category: 'new category',
        },
        loading: true
      }

      const tree = renderer.create(<AddCategory {...props} />).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('actions', () => {
    it('should be able to dispatch submit form actions', () => {
      const dispatch = jest.fn()

      const props = mapDispatchToProps(dispatch)
      props.addCategory()

      expect(dispatch).toBeCalledWith(submitForm(
        ADD_CATEGORY,
        '/api/category',
        handleResponse
      ))
    })
  })
})
