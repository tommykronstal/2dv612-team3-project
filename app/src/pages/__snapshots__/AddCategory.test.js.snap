// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCategory component renders correctly, initial 1`] = `
<section
  className="sc-bdVaJa cgbLUC"
>
  <form
    className="sc-bwzfXH gJIwTT"
    onSubmit={[Function]}
  >
    <h1
      className="sc-htpNat cfPrNC"
    >
      Add a new category
    </h1>
    <div
      className="sc-EHOje cprGDq"
    >
      <input
        className="sc-bxivhb hkzSDr"
        name="categoryName"
        onChange={[Function]}
        type="text"
        value=""
      />
      <label
        className="sc-ifAKCX crRGrj"
        value=""
      >
        Category Name
      </label>
    </div>
    <button
      className="sc-gzVnrw jmsTpF"
      disabled={undefined}
      onClick={[Function]}
    >
      Add
    </button>
  </form>
</section>
`;

exports[`AddCategory component renders correctly, with data 1`] = `
<section
  className="sc-bdVaJa cgbLUC"
>
  <form
    className="sc-bwzfXH gJIwTT"
    onSubmit={[Function]}
  >
    <h1
      className="sc-htpNat cfPrNC"
    >
      Add a new category
    </h1>
    <div
      className="sc-EHOje cprGDq"
    >
      <input
        className="sc-bxivhb hkzSDr"
        name="categoryName"
        onChange={[Function]}
        type="text"
        value=""
      />
      <label
        className="sc-ifAKCX crRGrj"
        value=""
      >
        Category Name
      </label>
    </div>
    <button
      className="sc-gzVnrw jmsTpF"
      disabled={undefined}
      onClick={[Function]}
    >
      Add
    </button>
  </form>
</section>
`;
